Data source: Fixes saving a data source without failure when URL doesn't specify a protocol

aknuds1 opened the pull request:
**What this PR does / why we need it**:
Modify data source URL validation to handle URLs that don't specify the protocol, f.ex. "127.0.0.1:5432". 

I've tested this change with Loki and Postgres data sources, and both work fine (I could verify that the bug in #24387, with specifying e.g. "127.0.0.1:5433" is fixed).

**Which issue(s) this PR fixes**:

Fixes #24387.

**Special notes for your reviewer**:
I don't feel as if it's the cleanest fix, but I'm not sure how else we might solve this, since a URL is inherently ambiguous without the protocol part (f.ex. "http://").

aknuds1 committed:
Data sources: Don't fail if URL doesn't specify protocol

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 commented on pkg/api/pluginproxy/ds_proxy.go:
```diff
@@ -83,10 +84,19 @@ func (lw *logWrapper) Write(p []byte) (n int, err error) {
+	u := ds.Url
+	// Make sure the URL starts with a protocol specifier, so parsing is unambiguous
+	if !reURL.MatchString(u) {
+		u = fmt.Sprintf("http://%s", u)
```
Could it cause problems anywhere that the URL's protocol is set to HTTP as a fallback?

bergquist replied:
I _think_ it should be fine as long as the regexp works as expected.

papagian replied:
I was wondering almost the same; is there any case that the fallback protocol should be `https` instead of `http`?
At least there could be a debug message that the silent implied protocol is `http`?

aknuds1 replied:
@papagian Would it help in any case to specify https? I am actually hoping that the protocol part of the URL won't be applied anywhere, since chances are it's incorrect. For example, I tested with Postgres, and it doesn't apply the protocol part of the URL (which would fail if it did).

Regarding https, I think it's less safe, since many servers might not support it and if you use http they will typically redirect you to https.

aknuds1 replied:
I've added debug logging BTW @papagian, please have a look.

papagian replied:
Honestly I can't tell for sure because I'm not aware of all the possible cases.
I think that the debug message should be enough for now.

I would ideally prefer also a unified `validateUrl` to be used by both the `api` and the `pluginproxy`. What do you think?

aknuds1 replied:
@papagian I can't share `validateURL` as is, but I can try to extract a common one.

papagian replied:
not a blocker anyway

bergquist commented on pkg/api/datasources.go:
```diff
@@ -129,8 +130,16 @@ func DeleteDataSourceByName(c *models.ReqContext) Response {
 func validateURL(u string) Response {
 	if u != "" {
+		// Make sure the URL starts with a protocol specifier, so parsing is unambiguous
+		if !reURL.MatchString(u) {
```
It would be nice with some tests for this. I always find it easier to understand regexp patterns if I can read tests :)

aknuds1 replied:
@bergquist Gotcha, I will add a test or two.

aknuds1 committed:
Debug log when prepending protocol to data source URLs

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Add tests

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Refactor data source URL validation

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into bugfix/fix-23661

aknuds1 committed:
Improve docs

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 commented:
@bergquist @papagian I added a collection of tests for adding and updating data sources, plus I refactored the URL validation into a shared function. Please have a look.

aknuds1 committed:
Remove unused var

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

bergquist commented on pkg/api/datasources_test.go:
```diff
@@ -56,3 +57,107 @@ func TestDataSourcesProxy(t *testing.T) {
+	defer bus.ClearBusHandlers()
+
+	sc := setupScenarioContext("/api/datasources/1234")
+	// TODO: Make this an argument to setupScenarioContext
```
I guess your goal is to solve these TODO's in another PR?

aknuds1 replied:
@bergquist Yes, it's a larger operation (as part of dropping goconvey), and I think best done separately.

bergquist reviewed (approved):
LGTM
